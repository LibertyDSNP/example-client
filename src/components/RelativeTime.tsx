import React from "react";

const getMonthName = (monthNum: number) => {
  const months = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sept",
    "Oct",
    "Nov",
    "Dec",
  ];
  return months[monthNum];
};

const getRelativeTime = (timestamp: number) => {
  const currentTime = new Date();
  const postTime = new Date(timestamp * 1000);
  const secondsPast = (currentTime.getTime() - postTime.getTime()) / 1000;
  if (secondsPast < 60) {
    return Math.floor(secondsPast) + "s •";
  }
  if (secondsPast < 3600) {
    return Math.floor(secondsPast / 60) + "m •";
  }
  if (secondsPast <= 86400) {
    return Math.floor(secondsPast / 3600) + "h •";
  }
  if (secondsPast > 86400) {
    const day = postTime.getDate();
    const month = getMonthName(postTime.getMonth());
    const year =
      postTime.getFullYear() === currentTime.getFullYear()
        ? ""
        : `, ${postTime.getFullYear()}`;
    return `${month} ${day}${year} •`;
  }
};

interface RelativeTimeProps {
  timestamp: number;
  postStyle: boolean;
}

const RelativeTime = ({
  timestamp,
  postStyle,
}: RelativeTimeProps): JSX.Element => {
  return (
    <p className={`RelativeTime ${postStyle ? "RelativeTime--post" : ""}`}>
      {getRelativeTime(timestamp)}
    </p>
  );
};

export default RelativeTime;
