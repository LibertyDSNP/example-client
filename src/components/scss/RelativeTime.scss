.RelativeTime {
  margin: 0;
  font-size: 13px;
}

.RelativeTime--post {
  font-weight: 500;
  align-items: flex-end;
}
