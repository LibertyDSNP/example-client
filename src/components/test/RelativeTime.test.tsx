import React from "react";
import renderer from "react-test-renderer";
import RelativeTime from "../RelativeTime";

describe("relative time", () => {
  it("renders without crashing", () => {
    const announcement = { published: new Date().toISOString() };
    expect(() => {
      renderer.create(
        <RelativeTime published={announcement.published} postStyle={true} />
      );
    }).not.toThrow();
  });
  it("renders seconds as expected", () => {
    const currentTime = new Date();
    const announcement = {
      published: new Date(
        currentTime.setSeconds(currentTime.getSeconds() - 10)
      ).toISOString(),
    };
    const rendered = renderer.create(
      <RelativeTime published={announcement.published} postStyle={true} />
    );
    expect(rendered).toMatchSnapshot();
  });
  it("renders minutes as expected", () => {
    const currentTime = new Date();
    const announcement = {
      published: new Date(
        currentTime.setMinutes(currentTime.getMinutes() - 1)
      ).toISOString(),
    };
    const rendered = renderer.create(
      <RelativeTime published={announcement.published} postStyle={true} />
    );
    expect(rendered).toMatchSnapshot();
  });
  it("renders hours as expected", () => {
    const currentTime = new Date();
    const announcement = {
      published: new Date(
        currentTime.setHours(currentTime.getHours() - 2)
      ).toISOString(),
    };
    const rendered = renderer.create(
      <RelativeTime published={announcement.published} postStyle={true} />
    );
    expect(rendered).toMatchSnapshot();
  });
  it("renders years as expected", () => {
    const currentTime = new Date("2018-11-18").toISOString();
    const announcement = {
      published: currentTime,
    };
    const rendered = renderer.create(
      <RelativeTime published={announcement.published} postStyle={true} />
    );
    expect(rendered).toMatchSnapshot();
  });
});
