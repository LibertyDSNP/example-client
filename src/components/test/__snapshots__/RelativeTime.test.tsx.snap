// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relative time renders hours as expected 1`] = `
<p
  className="RelativeTime RelativeTime--post"
>
  2h
</p>
`;

exports[`relative time renders minutes as expected 1`] = `
<p
  className="RelativeTime RelativeTime--post"
>
  1m
</p>
`;

exports[`relative time renders seconds as expected 1`] = `
<p
  className="RelativeTime RelativeTime--post"
>
  10s
</p>
`;

exports[`relative time renders years as expected 1`] = `
<p
  className="RelativeTime RelativeTime--post"
>
  Dec 18, 2018
</p>
`;
